### ISSUE TITLE
Suggestion: Infer literal types for properties of "frozen" object

### ISSUE THREAD

**Issue by levenleven** [Microsoft/TypeScript#29603]
## Search Terms

Object.freeze, literal types

## Suggestion

Infer literal types for properties of "frozen" object. See example below

## Use Cases

Moving project from flow to typescript. Common pattern to define enums in flow is freezing an object and using union of a value types as the enum type:
```ts
const foo = Object.freeze({ a: 'a', b: 'b' });
type FooValue = $Values<typeof foo>; // --> 'a' | 'b'
```
Currently same pattern in typescript will result in `string`.

## Examples

```ts
const foo = Object.freeze({ a: 'a', b: 'b' });
type Foo = typeof foo; // Expected: { readonly a: 'a'; readonly b: 'b'; } Actual: { readonly a: string; readonly b: string; }
type FooValue = Foo[keyof Foo] // Expected: 'a' | 'b' Actual: 'string'
```

## Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code (**Not sure about that**)
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by fatcerberus** [comment]
Do note that TypeScript has native strongly-typed enums built in:
https://www.typescriptlang.org/docs/handbook/enums.html#main-content

Something like this might be nice to have when typechecking JS code (`--checkJs`) though.

**Comment by levenleven** [comment]
Sure, enums are just example. Once object frozen, its values can't be changed, so it makes sense to infer their types as lietrals

**Comment by NN---** [comment]
I guess this one will solve it once merged  #29510
